### PR TITLE
QUIC LCIDM: Fix use of LHASH, clarify docs

### PR DESCRIPTION
**Pull request by hlandau**
- QUIC LCIDM: Fix usage of LHASH
- LHASH: Fix documentation for doall-delete hazards

I have done a quick look at the rest of the codebase and I do not think we have this bug anywhere else, as far as I can tell.

### CODE REVIEW

**Review by mattcaswell — APPROVED** [review]

**Comment by hlandau** [comment]
Agreed. This won't cherry-pick cleanly but adding the tags so I don't forget to open a new PR.

File: doc/man3/OPENSSL_LH_COMPFUNC.pod [hunk 1]
```diff
@@ -187,6 +178,23 @@ that is provided by the caller):
...
+callback passed to B<lh_I<TYPE>_doall>() or B<lh_I<TYPE>_doall_arg>(). The
+reason for this is that deleting an item from the hash table may result in the
+hash table being contracted to a smaller size and rehashed.
+B<lh_I<TYPE>_doall>() and B<lh_I<TYPE>_doall_arg>() are unsafe and will exhibit
+undefined behaviour under these conditions, as these functions assume the hash
+table size and bucket pointers do not change during the call.
+
+If it is desired to use B<lh_I<TYPE>_doall>() or B<lh_I<TYPE>_doall_arg>() with
+B<lh_I<TYPE>_delete>(), it is essential that you call
+B<lh_I<TYPE>_set_down_load>() with a I<down_load> argument of 0 first. This
```
Comment (kroeckx) [inline doc/man3/OPENSSL_LH_COMPFUNC.pod:191]: lh_I<TYPE>_set_down_load seems to de undocumented.

**Comment by hlandau** [comment]
Updated to document down_load

**Review by t8m — APPROVED** [review]

**Review by mattcaswell — APPROVED** [review]

**Comment by hlandau** [comment]
Merged. See https://github.com/openssl/openssl/pull/23032 for backports.

**Pull request closed without merging**